File: tutorials/workflow/csharp/challenges-tips/DeterministicWorkflow.cs
using System;
using Dapr.Workflow;

namespace WorkflowApp;

internal sealed class NonDeterministicWorkflow : Workflow<string, string>
{
    public override async Task<string> RunAsync(WorkflowContext context, string orderItem)
    {
        // Do not use non-deterministic operations in a workflow!
        // These operations will create a new value every time the
        // workflow is replayed.
        var orderId = Guid.NewGuid().ToString();
        var orderDate = DateTime.UtcNow;

        var idResult = await context.CallActivityAsync<string>(
            "SubmitId",
            orderId);

        await context.CallActivityAsync<string>(
            "SubmitDate",
            orderDate);

        return idResult;
    }
}

internal sealed class DeterministicWorkflow : Workflow<string, string>
{
    public override async Task<string> RunAsync(WorkflowContext context, string orderItem)
    {
        // Do use these deterministic methods and properties on the WorkflowContext instead.
        // These operations create the same value when the workflow is replayed.
        var orderId = context.NewGuid().ToString();
        var orderDate = context.CurrentUtcDateTime;

        var idResult = await context.CallActivityAsync<string>(
            "SubmitId",
            orderId);

        await context.CallActivityAsync<string>(
            "SubmitDate",
            orderDate);

        return idResult;
    }
}

File: tutorials/workflow/java/challenges-tips/DeterministicWorkflow.java
@Component
public class NonDeterministicWorkflow implements Workflow {
  @Override
  public WorkflowStub create() {
    return ctx -> {
      ctx.getLogger().info("Starting Workflow: " + ctx.getName());

      var orderItem = ctx.getInput(String.class);

      // Do not use non-deterministic operations in a workflow!
      // These operations will create a new value every time the
      // workflow is replayed.
      var orderId = UUID.randomUUID();;
      var orderDate = Instant.now();

      String idResult = ctx.callActivity(SubmitIdActivity.class.getName(), orderId, String.class).await();
      String dateResult = ctx.callActivity(SubmitDateActivity.class.getName(), orderDate, String.class).await();

      ctx.complete(idResult);
    };
  }
}

@Component
public class DeterministicWorkflow implements Workflow {
  @Override
  public WorkflowStub create() {
    return ctx -> {
      ctx.getLogger().info("Starting Workflow: " + ctx.getName());

      var orderItem = ctx.getInput(String.class);

      // Do use these deterministic methods and properties on the WorkflowContext instead.
      // These operations create the same value when the workflow is replayed.
      var orderId = ctx.newUuid();;
      var orderDate = ctx.getCurrentInstant();

      String idResult = ctx.callActivity(SubmitIdActivity.class.getName(), orderId, String.class).await();
      String dateResult = ctx.callActivity(SubmitDateActivity.class.getName(), orderDate, String.class).await();

      ctx.complete(idResult);
    };
  }
}